internal/controller/nodetemplate_controller.go

```diff
@@ -0,0 +1,390 @@
+package controller
+
+import (
+	"context"
+	"reflect"
+	"slices"
+	"time"
+
+	nyallocatorv1 "github.com/cybozu-go/nyallocator/api/v1"
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const (
+	SpareTaintKey                 = "node.cybozu.io/spare"
+	SpareRoleLabelKey             = "node-role.kubernetes.io/spare"
+	NodeTemplateReferenceLabelKey = "nyallocator.cybozu.io/node-template"
+	ZoneLabelKey                  = "topology.kubernetes.io/zone"
+	NodeTemplateFinalizer         = "nyallocator.cybozu.io/finalizer"
+)
+
+type NodeTemplateReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+}
+type NodeReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+// +kubebuilder:rbac:groups=nyallocator.cybozu.io,resources=nodetemplates,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=nyallocator.cybozu.io,resources=nodetemplates/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=nyallocator.cybozu.io,resources=nodetemplates/finalizers,verbs=update
+// +kubebuilder:rbac:groups=core,resources=nodes,verbs=get;list;watch;update;patch
+
+func (r *NodeTemplateReconciler) Reconcile(ctx context.Context, req ctrl.Request) (result ctrl.Result, err error) {
```

Comment:
Named return is not used in this function.
```suggestion
func (r *NodeTemplateReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
```

Reply:
It's used when status update in defer fails.
https://github.com/cybozu-go/nyallocator/blob/47365fc009d801d12e86295984d7b8af40f6ee88/internal/controller/nodetemplate_controller.go#L100

Reply:
Oh, I missed that.